### PR TITLE
fix(animations): normalize final styles in buildStyles

### PR DESCRIPTION
**Pull request by dario-piotrowicz**
the final styles created in buildStyles lack normalization, meaning that pixel values remain as numbers (without "px") and so such properties fail to be correctly set/applied

Example: "width: 300" would be applied as "width": "300" (and thus ignored) instead of the correct "width": "300px"

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?
At the end of animations numerical values are not normalized thus they are not saved correctly

Example with width, notice how the animation widths are ignored at the end:
width-before

[stackblitz](https://stackblitz.com/edit/angular-animation-end-width-bug)

Issue Number: https://github.com/angular/angular/issues/24698

## What is the new behavior?

numerical values get normalized at animations end thus fixing the issue

Example after fix:
width-after

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/animations/browser/src/util.ts [hunk 1]
```diff
@@ -159,14 +160,18 @@ function writeStyleAttribute(element: any) {
...
+    normalizer?: AnimationStyleNormalizer) {
   if (element['style']) {
     Object.keys(styles).forEach(prop => {
       const camelProp = dashCaseToCamelCase(prop);
       if (formerStyles && !formerStyles.hasOwnProperty(prop)) {
         formerStyles[prop] = element.style[camelProp];
       }
-      element.style[camelProp] = styles[prop];
+      const rawValue = styles[prop];
+      const value = normalizer?.normalizeStyleValue(prop, prop, rawValue, []) ?? rawValue;
```
Comment (JoostK) [inline packages/animations/browser/src/util.ts:173]: Since `normalizeStyleValue` does not return a nullable value, the `??` fallback is only here because of the usage of `.?` for `normalizer`. In that case a ternary expression would more closely indicate that, plus it generates less code as downleveling `??` and `.?` expands into multiple ternaries.

Comment (JoostK) [reply packages/animations/browser/src/util.ts:173]: More importantly is the question of whether `normalizer` should really be optional. Why is the normalizer only needed when `setStyles` is called from the `onDestroy` callback, but not in ~10 other calls to `setStyles`? That sounds inconsistent to me.

**Review by JoostK — COMMENTED** [review]
Heya, thank you for taking a look at this. The animations package is not a simple one to get your hands dirty on.

It would be great if this could have a test. I think the best place for such a test would be in the integration suite located in `packages/core/test/animation`. I am also a bit concerned about the normalizer being optional.

File: packages/animations/browser/src/util.ts (hunk 1 above)
Comment (dario-piotrowicz) [inline packages/animations/browser/src/util.ts:173]: having to call `normalizeStyleValue` with `prop`, `prop`, `rowValue` and `[]` is a bit annoying

but I noticed that there are already calls to the method with the first 2 arguments equal, so I guess that is not terrible

about passing `[]`, that is an array for string errors, I am just passing a new array and ignoring it as I think that there cannot be (new) errors, because if the animation is coming to an end the previous `setStyles` calls have already performed the checks

---

I thought of creating a new `AnimationStyleNormalizer` method which would only take the `prop` and `rawValue` (and would call `normalizeStyleValue` the same as I am doing here) but I figured I should not mess around to much with the existing classes and that this is preferable

File: packages/animations/browser/src/util.ts [hunk 2]
```diff
@@ -159,14 +160,18 @@ function writeStyleAttribute(element: any) {
   element.setAttribute('style', styleAttrValue);
 }
 
-export function setStyles(element: any, styles: ɵStyleData, formerStyles?: {[key: string]: any}) {
+export function setStyles(
+    element: any, styles: ɵStyleData, formerStyles?: {[key: string]: any}|null,
```
Comment (dario-piotrowicz) [inline packages/animations/browser/src/util.ts:164]: I added the `|null` as I wanted to pass that in the `setStyles(rootElement, instruction.toStyles, null, this._normalizer)` call

I could have just done `setStyles(rootElement, instruction.toStyles, undefined, this._normalizer)`, but using `undefined` that way just looks wrong to me

File: packages/animations/browser/src/util.ts (hunk 1 above)
Comment (dario-piotrowicz) [reply packages/animations/browser/src/util.ts:173]: about the use of `??` and `.?` I already actually thought of that and wanted to put a comment
(this version seems quite clean to me but yeah, not as performant as others)

--- 

About the `normalizer` being options, I must admit that was a very naive think of me, you're totally right that it should not be, I will amend this! :slightly_smiling_face: 

(the thing is that the bug as only related to the end of animations so I focused on that, but yeah, most likely having the values not normalized in `setStyles` would effect other aspects of animations)

**Comment by dario-piotrowicz** [comment]
> Heya, thank you for taking a look at this. The animations package is not a simple one to get your hands dirty on.
> 
> It would be great if this could have a test. I think the best place for such a test would be in the integration suite located in `packages/core/test/animation`. I am also a bit concerned about the normalizer being optional.

Thank you very much!!! :smiley: 

About the test, consider it done! :wink: 

Also the normalizer should definitely not be optional! you're totally right! I'm going to amend it :slightly_smiling_face:

File: packages/animations/browser/src/util.ts (hunk 2 above)
Comment (dario-piotrowicz) [reply packages/animations/browser/src/util.ts:164]: never mind this comment, I am going to make the `normalizer` not optional and thus there want be a need to pass `null` for the `formerStyles`

File: packages/animations/browser/src/util.ts [hunk 3]
```diff
@@ -159,14 +160,16 @@ function writeStyleAttribute(element: any) {
...
+    element: any, styles: ɵStyleData, normalizer: AnimationStyleNormalizer,
+    formerStyles?: {[key: string]: any}) {
   if (element['style']) {
     Object.keys(styles).forEach(prop => {
       const camelProp = dashCaseToCamelCase(prop);
       if (formerStyles && !formerStyles.hasOwnProperty(prop)) {
         formerStyles[prop] = element.style[camelProp];
       }
-      element.style[camelProp] = styles[prop];
+      element.style[camelProp] = normalizer.normalizeStyleValue(prop, prop, styles[prop], []);
```
Comment (dario-piotrowicz) [inline packages/animations/browser/src/util.ts:172]: here I think I will need to pass a proper `errors` array and check the result

(as in: https://github.com/angular/angular/blob/c9b47f21822343286d056fd70aec882acc0ae788/packages/animations/browser/src/render/shared.ts#L82)

File: packages/core/test/animation/animations_with_web_animations_integration_spec.ts [hunk 4]
```diff
@@ -512,6 +512,48 @@ describe('animation integration tests using web animations', function() {
        expect(elm.style.getPropertyValue('display')).toEqual('inline-block');
        expect(elm.style.getPropertyValue('position')).toEqual('fixed');
      });
+
+  it('should set normalized style property values on animation end', () => {
```
Comment (dario-piotrowicz) [inline packages/core/test/animation/animations_with_web_animations_integration_spec.ts:516]: @JoostK hopefully this is sort of what you had in mind? :slightly_smiling_face:

File: packages/animations/browser/src/util.ts [hunk 5]
```diff
@@ -159,14 +160,18 @@ function writeStyleAttribute(element: any) {
...
   if (element['style']) {
     Object.keys(styles).forEach(prop => {
       const camelProp = dashCaseToCamelCase(prop);
       if (formerStyles && !formerStyles.hasOwnProperty(prop)) {
         formerStyles[prop] = element.style[camelProp];
       }
-      element.style[camelProp] = styles[prop];
+      const normalizedProp = normalizer.normalizePropertyName(prop, []);
+      element.style[camelProp] =
+          normalizer.normalizeStyleValue(prop, normalizedProp, styles[prop], []);
```
Comment (dario-piotrowicz) [inline packages/animations/browser/src/util.ts:174]: @JoostK sorry for bothering you :cry: 

Could you advise on what to do with the errors here? 
(I am currently ignoring them since I just pass empty one-time arrays)

Should I gather the errors and throw as in: https://github.com/angular/angular/blob/c9b47f21822343286d056fd70aec882acc0ae788/packages/animations/browser/src/render/shared.ts#L82 ?

Comment (JoostK) [reply packages/animations/browser/src/util.ts:174]: That is a very good question. There is currently one error reported for the case where a string without unit is provided for a dimensional property. I think reporting that is reasonable, however that would introduce a breaking change and limits this change from landing in patch.

Therefore, I think it's reasonable to ignore the errors for now such that the behavior is not affected. We can later revisit this choice in a major version if needed. It may depend a bit on how breaky it turns out in Google.

File: packages/core/test/animation/animations_with_web_animations_integration_spec.ts (hunk 4 above)
Comment (JoostK) [reply packages/core/test/animation/animations_with_web_animations_integration_spec.ts:516]: I like it!

File: packages/animations/browser/src/util.ts (hunk 5 above)
Comment (JoostK) [reply packages/animations/browser/src/util.ts:174]: We may want to also have a test for the behavior we choose here; i.e. that a string without number bound to a dimension property applies the string verbatim without an error.

**Review by JoostK — COMMENTED** [review]
I think this looks good now. I'm marking for presubmit to see how this behaves in Google.

File: packages/core/test/animation/animations_with_web_animations_integration_spec.ts (hunk 4 above)
Comment (dario-piotrowicz) [reply packages/core/test/animation/animations_with_web_animations_integration_spec.ts:516]: thanks!!!! :smile:  :smile:  :smile:

File: packages/animations/browser/src/util.ts (hunk 5 above)
Comment (dario-piotrowicz) [reply packages/animations/browser/src/util.ts:174]: About the breaking change thing, I understand and yeah I would really like it to just get this fixed now and add the error handling later :slightly_smiling_face: 

would you mind if I add a comment there? 
(something like: `// TODO: (try to) add error handling before new major release (introduces a breaking change)`)

Comment (dario-piotrowicz) [reply packages/animations/browser/src/util.ts:174]: About the test, I actually tried to implement it so that but I actually think that the animations don't past a certain point if they don't conform to the standards

to be more clear, I tried to change my test and modified 300 to "300", when running it I got:
```
Error: Unable to animate due to the following errors:
         - Please provide a CSS unit value for width:300
```         
         
So I looked into it and it's the error thrown in `normalizeKeyframes` (the same I linked in my comment), that is called as part of the `TransitionAnimationEngine`'s `flush` method

So if I understood correctly, the `flush` happens still as part of the animation before the `destroy` to which I am mainly hooking into, so if there were actual errors in the values they would have been caught by `flush`, so I think that my ignoring those errors is actually correct as there may not be errors at this point, because if there were errors they would have already been caught and an error would have been thrown.....

at least that's my understanding of it (if you want I can keep looking into it and try to get more confidence in this)

does this checks out with you? what do you think?

Comment (JoostK) [reply packages/animations/browser/src/util.ts:174]: I see, thanks for checking. Based on your comment, I realize that the styles do get normalized at some point, but apparently not for all usages. That is an inconsistency in how the styles are interpreted, which is what you are addressing by normalizing the values as they are written.

I was interested to know where the `instruction.toStyles` came from for which the normalization was necessary. These styles originate here:

https://github.com/angular/angular/blob/94e3d3ab70402629fb5882523bc2506d9551dac0/packages/animations/browser/src/dsl/animation_transition_factory.ts#L46-L49

Which calls into the following:

https://github.com/angular/angular/blob/94e3d3ab70402629fb5882523bc2506d9551dac0/packages/animations/browser/src/dsl/animation_transition_factory.ts#L104-L126

I feel like that would actually be the appropriate place to do the normalization as it's still during transaction building where errors can be reported, rather than during write (which creates the unfortunate situation where errors are no longer expected, hence the questions about how to deal with the errors). I think you can get the normalizer in `buildStyles` by adding it to `AnimationTransitionFactory.build`, `AnimationTransitionFactory.buildStyles` and `AnimationStateStyles.buildStyles`. It should then hopefully be possible to normalize the value that is written to `finalStyles`.

Comment (dario-piotrowicz) [reply packages/animations/browser/src/util.ts:174]: Oh gosh, I am so sorry you had to look into things so much because of me :cry: 

I see what you mean, I will check it and try to place the normalization where it makes more sense :slightly_smiling_face: 

( again sorry for not having looked at this well enough :sob: )

Comment (JoostK) [reply packages/animations/browser/src/util.ts:174]: Please, don't be sorry! This is how both of us get to learn something new :)

**Comment by dario-piotrowicz** [comment]
@JoostK of course your comment was spot on and I put the normalization in the `AnimationStateStyles`'s `buildStyles` method :slightly_smiling_face: 

Hopefully it should be all good now and I am not ignoring the error messages

I played around with it and all seems to be working fine, both correct and incorrect scenarios, camel cased styles and not etc...

File: packages/animations/browser/test/render/transition_animation_engine_spec.ts [hunk 6]
```diff
@@ -20,7 +20,7 @@ const DEFAULT_NAMESPACE_ID = 'id';
 (function() {
 const driver = new MockAnimationDriver();
 
-// these tests are only mean't to be run within the DOM
+// these tests are only meant to be run within the DOM
```
Comment (dario-piotrowicz) [inline packages/animations/browser/test/render/transition_animation_engine_spec.ts:23]: this is kind of unrelated, please let me know if I should put this into a separate commit or it's fine here :slightly_smiling_face:

**Comment by dario-piotrowicz** [comment]
@JoostK above my changes I noticed the `interpolateParams` call and I looked into it as I was curious about it and it's relationship with the normalization

I saw something interesting there, basically there the normalization breaks for numberical values, what I mean is the for example if I try to pass the number `20` as `font-size` I get an error: 
Screenshot at 2021-07-10 15-27-18

I believe that must be related to the template params parsing, the param number must be converted to a string and that's why the normalization fails (the number `20` gets normalized to `"20px"`, but if you pass `"20"` that's just rightfully considered a wrong css value).

What's even more curious to me, is that even a numerical value provided in the default params (not in the template) also causes the same error (that could have been done on purpose to make the template and default params behaviours consistent).

( By the way, I did check and this is not caused by my changes, it already happens in the current angular version :stuck_out_tongue_winking_eye:  )

I think that that is unrelated here and that I should not touch the interpolated values in this PR but may be interesting to look into next and see if I can do anything about it (I guess it all depends on whether the template values can be correctly parsed as numbers)

Please let me know what you think about it :slightly_smiling_face:

**Comment by dario-piotrowicz** [comment]
@JoostK? :cry:

**Comment by JoostK** [comment]
Heya, sorry it took me a while to respond. I think I understand what you mean in your last comment about parameter normalization and I agree that it's a different issue, so no need to worry about that in this PR. I think I agree with your analysis that numeric parameters should not cause normalization failures due to a hidden conversion to strings.

**Comment by dario-piotrowicz** [comment]
> Heya, sorry it took me a while to respond. I think I understand what you mean in your last comment about parameter normalization and I agree that it's a different issue, so no need to worry about that in this PR. I think I agree with your analysis that numeric parameters should not cause normalization failures due to a hidden conversion to strings.

Ah no worries, sorry for bothering you, I was just afraid you forgot about this PR :sweat_smile:
(I really really hope I didn't bother you :sweat: )

Thanks for the confirmation on the other thing, I will keep it in mind then :slightly_smiling_face:

File: packages/animations/browser/src/dsl/animation_transition_factory.ts [hunk 7]
```diff
@@ -118,6 +122,8 @@ export class AnimationStateStyles {
           if (val.length > 1) {
             val = interpolateParams(val, combinedParams, errors);
           }
+          const normalizedProp = normalizer.normalizePropertyName(prop, errors);
```
[the remark this answers is not in the document]
Comment (dario-piotrowicz) [reply packages/animations/browser/src/dsl/animation_transition_factory.ts:125]: actually I did check that and I remember I did not put `normalizedProp` in the `finalStyles` as that changed the final output of the styles building, so I just wanted to keep my changes to a minimum and just fix what's needed without changing behaviour, I'll have another look and will put a more detailed comment so you can advise further :slightly_smiling_face:

File: packages/animations/browser/src/dsl/animation_trigger.ts [hunk 8]
```diff
@@ -11,14 +11,16 @@ import {copyStyles, interpolateParams} from '../util';
...
 
 
 
 /**
  * @publicApi
  */
-export function buildTrigger(name: string, ast: TriggerAst): AnimationTrigger {
-  return new AnimationTrigger(name, ast);
+export function buildTrigger(
+    name: string, ast: TriggerAst, normalizer: AnimationStyleNormalizer): AnimationTrigger {
```
[the remark this answers is not in the document]
Comment (dario-piotrowicz) [reply packages/animations/browser/src/dsl/animation_trigger.ts:22]: ah wow! sorry I did not notice that!

I find this strange thought... it looks like a pretty low level function to be in the public API :confused: 

What can we do about this?  :disappointed:

File: packages/animations/browser/test/render/transition_animation_engine_spec.ts (hunk 6 above)
Comment (dario-piotrowicz) [reply packages/animations/browser/test/render/transition_animation_engine_spec.ts:23]: cool, I moved this to a separate commit then thanks! :slightly_smiling_face:

File: packages/animations/browser/src/dsl/animation_transition_factory.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (dario-piotrowicz) [reply packages/animations/browser/src/dsl/animation_transition_factory.ts:125]: ok I think I got what you were saying :sweat_smile: , the create instruction will actually be different

(
for example this is an instruction we get without using `normalizedProp` here:
(note the `toStyles`'s font size)
instructionWithout

and this is the instruction we get for the same animation if we do use the `normalizedProp` here:
instructionWith
)

so we do actually get a different instruction, but then the `fromStyles` and `toStyles` are always used by either `setStyles` or `eraseStyles` which normalize the prop, so the difference in the instructions really doesn't make any difference

right? :slightly_smiling_face:

[the remark this answers is not in the document]
Comment (JoostK) [reply packages/animations/browser/src/dsl/animation_transition_factory.ts:125]: Indeed. There's a slight difference if both casings are present but I don't think it makes an observable difference in the end.

File: packages/animations/browser/src/dsl/animation_trigger.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (JoostK) [reply packages/animations/browser/src/dsl/animation_trigger.ts:22]: I would add a commit that removes the `@publicApi` marker from `buildTrigger` and `AnimationTrigger` at the beginning of this PR.

File: packages/animations/browser/src/dsl/animation_transition_factory.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (dario-piotrowicz) [reply packages/animations/browser/src/dsl/animation_transition_factory.ts:125]: ah ok thanks! I made the change :slightly_smiling_face:

File: packages/animations/browser/src/dsl/animation_trigger.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (dario-piotrowicz) [reply packages/animations/browser/src/dsl/animation_trigger.ts:17]: Ok done thanks a lot! :slightly_smiling_face: 

___

In the comment message I've put a link to this comments thread so I just wanted to add some info here for anyone how could look into this:
- the `AnimationTrigger` is indeed not present in the [aio animations public api docs](https://angular.io/api/animations), there is an `AnimationTriggerMetadata` but it seems unrelated to the class
- in the codebase `AnimationTrigger` doesn't seem to be exported publicly (it does not surface to the animation's [`public_api.ts`](https://github.com/angular/angular/blob/master/packages/animations/public_api.ts))
- `buildTrigger` is also not present in the docs nor it seems to be exported publicly

File: packages/animations/browser/src/dsl/animation_trigger.ts [hunk 9]
```diff
@@ -7,29 +7,24 @@
  */
 import {AnimationMetadataType, ɵStyleData} from '@angular/animations';
 
-import {copyStyles, interpolateParams} from '../util';
-
-import {SequenceAst, StyleAst, TransitionAst, TriggerAst} from './animation_ast';
+import {SequenceAst, TransitionAst, TriggerAst} from './animation_ast';
```
Comment (dario-piotrowicz) [inline packages/animations/browser/src/dsl/animation_trigger.ts:10]: I removed some unused imports here, I hope this if fine :slightly_smiling_face:

File: packages/animations/browser/src/dsl/animation_trigger.ts [hunk 10]
```diff
@@ -15,17 +15,11 @@ import {AnimationStyleNormalizer} from './style_normalization/animation_style_no
 
 
 
-/**
```
Comment (JoostK) [inline packages/animations/browser/src/dsl/animation_trigger.ts:17]: I'd change the `fix(animations): remove publicApi annotations from triggers` commit to a `refactor(animations)`, as this change is not a user-facing fix (and does not need to go into the changelog)

Comment (dario-piotrowicz) [reply packages/animations/browser/src/dsl/animation_trigger.ts:17]: ah ok sorry :sweat: 

I'll change the other commit messages too then :slightly_smiling_face: 

thanks!!! :slightly_smiling_face:

**Review by JoostK — APPROVED** [review]

**Comment by JoostK** [comment]
I'm not an animations code owner so my approval is not sufficient. Hopefully @crisbeto can also have a look at this one.

**Comment by dario-piotrowicz** [comment]
> I'm not an animations code owner so my approval is not sufficient. Hopefully @crisbeto can also have a look at this one.

Awesome!!! Thank you so very much for all your help!!!! :heart_eyes: :smiling_face_with_three_hearts:

**Comment by AndrewKushnir** [comment]
FYI, I've started tests in [Google's codebase](https://test.corp.google.com/OCL:385334798:BASE:385334847:1626547608045:8f28cd22) (internal-only link) and will keep this thread updated once I get results.

File: packages/animations/browser/src/dsl/animation_transition_factory.ts [hunk 11]
```diff
@@ -101,7 +104,8 @@ function oneOrMoreTransitionsMatch(
 export class AnimationStateStyles {
   constructor(private styles: StyleAst, private defaultParams: {[key: string]: any}) {}
 
-  buildStyles(params: {[key: string]: any}, errors: string[]): ɵStyleData {
+  buildStyles(params: {[key: string]: any}, errors: string[], normalizer: AnimationStyleNormalizer):
```
Comment (crisbeto) [inline packages/animations/browser/src/dsl/animation_transition_factory.ts:107]: Would it make sense to pass in the normalizer as a parameter to the `AnimationStateStyles` constructor? I don't imagine it's going to change between calls.

**Review by crisbeto — APPROVED** [review]
Left one question, but otherwise LGTM.

File: packages/animations/browser/src/dsl/animation_transition_factory.ts (hunk 11 above)
Comment (dario-piotrowicz) [reply packages/animations/browser/src/dsl/animation_transition_factory.ts:107]: yeah... that's a very good point....

sorry I missed that, I'll amend it now :sweat:

**Comment by JoostK** [comment]
@dario-piotrowicz You seem to have amended the last commit, which was the refactor to remove the imports. I think the goal was to create a fixup commit and squash that into the second commit.

**Comment by dario-piotrowicz** [comment]
> @dario-piotrowicz You seem to have amended the last commit, which was the refactor to remove the imports. I think the goal was to create a fixup commit and squash that into the second commit.

lol you noticed that immediately! :see_no_evil: 

force of habit to amend the last commit... sorry :sweat_smile: , should be all good now

File: packages/animations/browser/src/dsl/animation_transition_factory.ts (hunk 11 above)
Comment (dario-piotrowicz) [reply packages/animations/browser/src/dsl/animation_transition_factory.ts]: all done thanks! :slightly_smiling_face:

**Comment by AndrewKushnir** [comment]
FYI tests in Google's codebase (including a global run) went well.

// cc @JoostK

File: packages/animations/browser/test/dsl/animation_trigger_spec.ts [hunk 12]
```diff
@@ -9,6 +9,7 @@
 import {animate, AnimationOptions, state, style, transition} from '@angular/animations';
 import {AnimationTransitionInstruction} from '@angular/animations/browser/src/dsl/animation_transition_instruction';
 import {AnimationTrigger} from '@angular/animations/browser/src/dsl/animation_trigger';
+import {NoopAnimationStyleNormalizer} from '../../src/dsl/style_normalization/animation_style_normalizer';
```
Comment (AndrewKushnir) [inline packages/animations/browser/test/dsl/animation_trigger_spec.ts:12]: The `NoopAnimationStyleNormalizer` looks unused.

**Review by AndrewKushnir — APPROVED** [review]
@dario-piotrowicz thanks for the fix 👍
The changes look good, just left 1 minor comment on seemingly unused import.

File: packages/animations/browser/test/dsl/animation_trigger_spec.ts (hunk 12 above)
Comment (dario-piotrowicz) [reply packages/animations/browser/test/dsl/animation_trigger_spec.ts]: sorry I forgot to remove it during the last refactoring :bow:

**Comment by AndrewKushnir** [comment]
@JoostK @crisbeto could you please have a final look and if everything is ok, we can probably add the PR to the merge queue?

**Comment by JoostK** [comment]
Let's do it! I was waiting for CI to go green.
